### ISSUE TITLE
Cannot use CustomData with Managed Disks

### ISSUE THREAD

**Issue by vincentdo** [Azure/azure-sdk-for-java#1628]
In the beta5 build, we are able to create a machine with a custom image and using the following code:

```java
azure.virtualMachines().define(pName)
            .withRegion(pRegion)
            .withExistingResourceGroup(pResourceGroupName)
            .withExistingPrimaryNetworkInterface(networkInterface)
            .withLinuxCustomImage(pImageId)
            .withRootUsername("dummyUsername")
            .withSsh(pSSHPublicKey)
            .withCustomData(pUserData)
            .withOSDiskName(pDiskName)
            .withSize(pAzureInstanceSize.getAzureInstanceTypeName())
            ::createAsync
```
This is no longer the case in v1.0.0 (the `withCustomData` chain is no longer available after `.withLinuxCustomImage`) and I believe this is a regression. There's no reason why we shouldn't be able to specify custom data on machine creation while using managed disks.

**Comment by jianghaolu** [comment]
@martinsawicki you have some context here?

**Comment by martinsawicki** [comment]
@anuchandy would know for sure, but he's on vacation. From what I can tell, `.withLinuxCustomImage()` takes you down the managed disk-only flow which for some reason does not offer the `.withCustomData()` option in v1.0.0. The `.withPopularLinuxImage()` path would take you down the Managed-or-Unmanaged disk flow ([`WithFromImageCreateOptionsManagedOrUnmanaged`](https://github.com/Azure/azure-sdk-for-java/blob/master/azure-mgmt-compute/src/main/java/com/microsoft/azure/management/compute/VirtualMachine.java#L878)), where `.withCustomData()` is available and we have a [test for that](https://github.com/Azure/azure-sdk-for-java/blob/master/azure/src/test/java/com/microsoft/azure/management/TestVirtualMachineCustomData.java#L64).

I wonder if `.withCustomData()` should have been part of the [`WithFromImageCreateOptionsManaged`](https://github.com/Azure/azure-sdk-for-java/blob/master/azure-mgmt-compute/src/main/java/com/microsoft/azure/management/compute/VirtualMachine.java#L884) interface instead (or additionally).

I'd wait for @anuchandy to determine this. In the meantime though, a temporary workaround could be to cast the output of that `.withSsh()` wither in the code sample above to `VirtualMachine.DefinitionStages.WithLinuxCreateManagedOrUnmanaged`:

```java
((VirtualMachine.DefinitionStages.WithLinuxCreateManagedOrUnmanaged) // Start of cast
azure.virtualMachines().define(pName)
    .withRegion(pRegion)
    .withExistingResourceGroup(pResourceGroupName)
    .withExistingPrimaryNetworkInterface(networkInterface)
    .withLinuxCustomImage(pImageId)
    .withRootUsername("dummyUsername")
    .withSsh(pSSHPublicKey)) // End of cast
    .withCustomData(pUserData)
    .withOSDiskName(pDiskName)
    .withSize(pAzureInstanceSize.getAzureInstanceTypeName())
    ::createAsync
```

This should be a safe cast because all these interfaces are implemented by one and the same impl class under the hood.

@vincentdo could you pls see if this successfully works around your challenge?

(If my hunch is correct, I think this should be fixable in a non-back compat breaking way soon.)

**Comment by vincentdo** [comment]
Using the cast as a work around seems to be working.

**Comment by martinsawicki** [comment]
Good to hear. We'll hold on to this issue until it is addressed in the next significant update (early June).

**Comment by anuchandy** [comment]
@martinsawicki Looks like it's a regression happened as part of the doc scrub we did. The definition used to be 

```java
        ....
        interface WithFromImageCreateOptionsManagedOrUnmanaged extends WithFromImageCreateOptionsManaged {
            WithFromImageCreateOptionsUnmanaged withUnmanagedDisks();
        }

        interface WithFromImageCreateOptionsManaged extends WithManagedCreate {
            WithFromImageCreateOptionsManaged withCustomData(String base64EncodedCustomData);
            WithFromImageCreateOptionsManaged withComputerName(String computerName);
        }
        ....
```

and after our doc clean up [here](https://github.com/Azure/azure-sdk-for-java/commit/fa2980ee9c207cde082ff9b9b4c4c5b1374376c6#diff-d658d729327a610762ae129df368ce3c) it become:

```java
        ...
        interface WithFromImageCreateOptionsManagedOrUnmanaged extends WithFromImageCreateOptionsManaged {
            WithFromImageCreateOptionsUnmanaged withUnmanagedDisks();
            WithFromImageCreateOptionsManaged withCustomData(String base64EncodedCustomData);
        }


        interface WithFromImageCreateOptionsManaged extends WithManagedCreate {
            WithFromImageCreateOptionsManaged withComputerName(String computerName);
        }
        ...
```

We just need to revert this change. I can send a PR.

**Comment by martinsawicki** [comment]
that is very weird, since I wasn't (intentionally) changing interface definitions, just javadocs. Could it have been a bad Git merge or diff?
In any case, sure, it looks like putting `withCustomData()` back on `WithFromImageCreateOptionsManaged` would not have any impact on back compat.

**Comment by anuchandy** [comment]
yes, putting it back is not a compact break. It seems a git merge issue.

**Comment by martinsawicki** [comment]
fixed for v1.1.0

**Comment by ignatev** [comment]
Hello. 
I can't use wtihCustomData() method in this snippet:
Azure java library version: com.microsoft.azure:azure:1.9.0
```
            Creatable<VirtualMachine> vm = azure.virtualMachines().define(vmType + "-" + i)
                    .withRegion(region)
                    .withExistingResourceGroup(resourceGroup.name())
                    .withExistingPrimaryNetwork(network)
                    .withSubnet("subnet")
                    .withPrimaryPrivateIPAddressStatic("10.100.10.1")
                    .withoutPrimaryPublicIPAddress()
                    .withLinuxCustomImage(vmImage)
                    .withRootUsername("devops")
                    .withRootPassword("devops")
                    .withSsh("ssh-rsa ")
                    .withSize(vmSize);
```
